Mathlib/Combinatorics/Graph/Basic.lean

```diff
@@ -0,0 +1,286 @@
+lemma Inc.inc_other (h : G.Inc e x) : G.Inc e h.other :=
+  h.inc₂_other.inc_right
+
+lemma Inc.eq_or_eq_or_eq_of_inc_of_inc (hx : G.Inc e x) (hy : G.Inc e y) (hz : G.Inc e z) :
+    x = y ∨ x = z ∨ y = z := by
+  by_contra! hcon
+  obtain ⟨x', hx'⟩ := hx
+  obtain rfl := hy.eq_of_inc₂_of_ne_left hx' hcon.1.symm
+  obtain rfl := hz.eq_of_inc₂_of_ne_left hx' hcon.2.1.symm
+  exact hcon.2.2 rfl
+
+/-- `G.IsLoopAt e x` means that both ends of `e` are equal to `x`. -/
+def IsLoopAt (G : Graph α β) (e : β) (x : α) : Prop := G.Inc₂ e x x
+
+@[simp]
+lemma inc₂_self_iff : G.Inc₂ e x x ↔ G.IsLoopAt e x := Iff.rfl
+
+lemma IsLoopAt.inc (h : G.IsLoopAt e x) : G.Inc e x :=
+  Inc₂.inc_left h
+
+lemma IsLoopAt.eq_of_inc (h : G.IsLoopAt e x) (h' : G.Inc e y) : x = y := by
+  obtain rfl | rfl := h'.eq_or_eq_of_inc₂ h <;> rfl
+
+@[simp]
+lemma IsLoopAt.edge_mem (h : G.IsLoopAt e x) : e ∈ G.E :=
+  h.inc.edge_mem
+
+@[simp]
+lemma IsLoopAt.vx_mem (h : G.IsLoopAt e x) : x ∈ G.V :=
+  h.inc.vx_mem
+
+/-- `G.IsNonloopAt e x` means that `e` is an edge from `x` to some `y ≠ x`. -/
+@[mk_iff]
+structure IsNonloopAt (G : Graph α β) (e : β) (x : α) : Prop where
+  inc : G.Inc e x
+  exists_inc₂_ne : ∃ y ≠ x, G.Inc₂ e x y
+
+@[simp]
+lemma IsNonloopAt.edge_mem (h : G.IsNonloopAt e x) : e ∈ G.E :=
+  h.inc.edge_mem
+
+@[simp]
+lemma IsNonloopAt.vx_mem (h : G.IsNonloopAt e x) : x ∈ G.V :=
+  h.inc.vx_mem
+
+lemma IsLoopAt.not_isNonloop_at (h : G.IsLoopAt e x) (y : α) : ¬ G.IsNonloopAt e y := by
+  intro h'
+  obtain ⟨z, hyz, hy⟩ := h'.exists_inc₂_ne
+  rw [← h.eq_of_inc hy.inc_left, ← h.eq_of_inc hy.inc_right] at hyz
+  exact hyz rfl
+
+lemma IsNonloopAt.not_isLoopAt (h : G.IsNonloopAt e x) (y : α) : ¬ G.IsLoopAt e y :=
+  fun h' ↦ h'.not_isNonloop_at x h
+
+lemma Inc.isLoopAt_or_isNonloopAt (h : G.Inc e x) : G.IsLoopAt e x ∨ G.IsNonloopAt e x := by
+  obtain ⟨y, hy⟩ := h
+  obtain rfl | hne := eq_or_ne x y
+  · exact .inl hy
+  exact .inr ⟨hy.inc_left, y, hne.symm, hy⟩
+
+/-! ### Adjacency -/
+
+/-- `G.Adj x y` means that `G` has an edge from `x` to `y`. -/
+def Adj (G : Graph α β) (x y : α) : Prop := ∃ e, G.Inc₂ e x y
+
+lemma Adj.symm (h : G.Adj x y) : G.Adj y x :=
+  ⟨_, h.choose_spec.symm⟩
+
+lemma adj_comm : G.Adj x y ↔ G.Adj y x :=
+  ⟨Adj.symm, Adj.symm⟩
+
+@[simp]
+lemma Adj.mem_left (h : G.Adj x y) : x ∈ G.V :=
+  h.choose_spec.vx_mem_left
+
+@[simp]
+lemma Adj.mem_right (h : G.Adj x y) : y ∈ G.V :=
+  h.symm.mem_left
+
+lemma Inc₂.adj (h : G.Inc₂ e x y) : G.Adj x y :=
+  ⟨e, h⟩
+
+/-! ### Extensionality -/
+
+/-- A constructor for `Graph` in which the edge set is inferred from the incidence predicate
+rather than supplied explicitly. -/
+@[simps]
+protected def mk' (V : Set α) (Inc₂ : β → α → α → Prop)
+    (inc₂_symm : ∀ ⦃e x y⦄, Inc₂ e x y → Inc₂ e y x)
+    (eq_or_eq_of_inc₂_of_inc₂ : ∀ ⦃e x y v w⦄, Inc₂ e x y → Inc₂ e v w → x = v ∨ x = w)
+    (vx_mem_left_of_inc₂ : ∀ ⦃e x y⦄, Inc₂ e x y → x ∈ V) : Graph α β where
+  V := V
+  E := {e | ∃ x y, Inc₂ e x y}
```

Comment:
If you reorder the fields in `Graph` itself, moving `E` later you could have `E : Set β := {e | ∃ x y, Inc₂ e x y}` to set this as a default value, without needing to have this separate constructor.

The `edge_mem_iff_exists_inc\_2` field could use `by intro; rfl` as a default value.

There's no harm in a separate constructor though. It's more flexible anyway.

Reply:
Maybe I'll stick with the separate constructor, then.